Sources/NIOCore/Codec.swift

```diff
@@ -740,12 +727,77 @@ extension ByteToMessageHandler: RemovableChannelHandler {
     }
 }
 
+/// A protocol for straightforward encoders which encode custom messages to `ByteBuffer`s.
+/// To add a `MessageToByteEncoder` to a `ChannelPipeline`, use
+/// `channel.pipeline.addHandler(MessageToByteHandler(myEncoder)`.
+public protocol MessageToByteEncoder {
+    associatedtype OutboundIn
+
+    /// Called once there is data to encode.
+    ///
+    /// - parameters:
+    ///     - data: The data to encode into a `ByteBuffer`.
+    ///     - out: The `ByteBuffer` into which we want to encode.
+    func encode(data: OutboundIn, out: inout ByteBuffer) throws
+
+    /// Called once there is data to encode and  the `MessageToByteDecoder` is about to leave
+    ///
+    /// - parameters:
+    ///     - data: The data to encode into a `ByteBuffer`.
+    ///     - out: The `ByteBuffer` into which we want to encode.
+    func encodeLast(data: OutboundIn, out: inout ByteBuffer) throws
+
+    /// Called once this `MessageToByteEncoder` is removed from the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ByteToMessageDecoder` belongs to.
+    mutating func encoderRemoved(context: ChannelHandlerContext)
+
+    /// Called when this `MessageToByteEncoder` is added to the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `MessageToByteEncoder` belongs to.
+    mutating func encoderAdded(context: ChannelHandlerContext)
+}
+
+extension MessageToByteEncoder {
+
+    public mutating func encoderRemoved(context: ChannelHandlerContext) {
+    }
+
+    public mutating func encoderAdded(context: ChannelHandlerContext) {
+    }
+
+    public func encodeLast(data: OutboundIn, out: inout ByteBuffer) throws {
+        try self.encode(data: data, out: &out)
+    }
+}
+
 /// A handler which turns a given `MessageToByteEncoder` into a `ChannelOutboundHandler` that can then be added to a
 /// `ChannelPipeline`.
 public final class MessageToByteHandler<Encoder: MessageToByteEncoder>: ChannelOutboundHandler {
     public typealias OutboundOut = ByteBuffer
     public typealias OutboundIn = Encoder.OutboundIn
 
+    private enum RemovalState {
+        /// Not added to any `ChannelPipeline` yet.
+        case notAddedToPipeline
+
+        /// No one tried to remove this handler.
+        case notBeingRemoved
+
+        /// The user-triggered removal has been started but isn't complete yet. This state will not be entered if the
+        /// removal is triggered by Channel teardown.
+        case removalStarted
+
+        /// The user-triggered removal is complete. This state will not be entered if the removal is triggered by
+        /// Channel teardown.
+        case removalCompleted
+
+        /// This handler has been removed from the pipeline.
+        case handlerRemovedCalled
+    }
```

Comment:
I don't think we need this complexity.

The complexity in `ByteToMessageEncoder` is present because of the fact that the encoder gets access to the `ChannelHandlerContext`, and so can trigger arbitrary re-entrant calls. `MessageToByteEncoder` does not allow this: you can only write into the `ByteBuffer`. So in this case, we can skip defending against re-entrancy.

As a result, we can remove almost all of this: removal can be done essentially immediately, with only the need to notify the encoder and give it an opportunity to write a final message. Does this make sense?

Reply:
Yeah that's mostly what I thought as there is not pending buffer. Will update ASAP.